### PR TITLE
test: add e2e test for create snapshot on external resource group

### LINKED ISSUE #533: add e2e test for create snapshot on external resource group
**Is your feature request related to a problem?/Why is this needed**

**Describe the solution you'd like in detail**

to prevent bug: https://github.com/kubernetes-sigs/azuredisk-csi-driver/issues/530

also need to add delete snapshot in e2e test

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Sakuralbj**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #533 

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

**Review by andyzhangx — APPROVED** [review]
/ok-to-test

File: test/e2e/testsuites/specs.go [hunk 1]
```diff
@@ -251,3 +256,24 @@ func CreateVolumeSnapshotClass(client restclientset.Interface, namespace *v1.Nam
 
 	return tvsc, tvsc.Cleanup
 }
+
+func CreateVolumeSnapshotClassWithExternalResourceGroup(client restclientset.Interface, namespace *v1.Namespace, csiDriver driver.VolumeSnapshotTestDriver, creds *credentials.Credentials) (*TestVolumeSnapshotClass, func()) {
```
Comment (andyzhangx) [inline test/e2e/testsuites/specs.go:260]: what about reuse `CreateVolumeSnapshotClass` func and create resource group outside

File: test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go [hunk 2]
```diff
@@ -53,9 +62,36 @@ func (t *DynamicallyProvisionedVolumeSnapshotTest) Run(client clientset.Interfac
...
+	framework.ExpectNoError(err, fmt.Sprintf("Error getting creds for AzurePublicCloud %v", err))
+	defer func() {
+		err := credentials.DeleteAzureCredentialFile()
+		framework.ExpectNoError(err)
+	}()
+	azureClient, err := azure.GetAzureClient(creds.Cloud, creds.SubscriptionID, creds.AADClientID, creds.TenantID, creds.AADClientSecret)
+	framework.ExpectNoError(err)
+
+	ctx := context.Background()
+	_, err = azureClient.EnsureResourceGroup(ctx, creds.ResourceGroup, creds.Location, nil)
```
Comment (andyzhangx) [inline test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:76]: so it's still create snapshot on current resource group?

Comment (Sakuralbj) [reply test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:76]: It wiil create a new resource group and convert it to volume snapshot class.

File: test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go [hunk 3]
```diff
@@ -53,9 +62,38 @@ func (t *DynamicallyProvisionedVolumeSnapshotTest) Run(client clientset.Interfac
 	ginkgo.By("checking that the pod's command exits with no error")
 	tpod.WaitForSuccess()
 
-	ginkgo.By("creating volume snapshot class")
+	ginkgo.By("Creating test resource group")
+	creds, err := credentials.CreateAzureCredentialFile(false)
+	ginkgo.By(creds.ResourceGroup)
+	creds.ResourceGroup = "azuredisk-csi-driver-test-e2e-test-external-rg"
```
Comment (andyzhangx) [inline test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:68]: should add a uuid here, otherwise there could be conflict by two PRs.

Comment (andyzhangx) [reply test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:68]: also print out the newly created rg by moving L65 to L69
```
ginkgo.By("Creating test resource group: %s", ...)
```

File: test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go [hunk 4]
```diff
@@ -53,9 +62,38 @@ func (t *DynamicallyProvisionedVolumeSnapshotTest) Run(client clientset.Interfac
...
 	tvsc, cleanup := CreateVolumeSnapshotClass(restclient, namespace, t.CSIDriver)
+	mp := map[string]string{
+		"resourceGroup": creds.ResourceGroup,
+	}
+	tvsc.volumeSnapshotClass.Parameters = mp
+	tvsc.Create()
 	defer cleanup()
+	defer func() {
+		// Only delete resource group the test created
+		if strings.HasPrefix(creds.ResourceGroup, credentials.ResourceGroupPrefix) {
```
Comment (andyzhangx) [inline test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:91]: so also need to change the rg detect condition here?

Comment (Sakuralbj) [reply test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:91]: I'll change it

File: test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go [hunk 5]
```diff
@@ -46,16 +56,51 @@ func (t *DynamicallyProvisionedVolumeSnapshotTest) Run(client clientset.Interfac
...
+	defer func() {
+		err := credentials.DeleteAzureCredentialFile()
+		framework.ExpectNoError(err)
+	}()
+
+	text := "Prow test resource group: " + creds.ResourceGroup
+	ginkgo.By(text)
+
+	//create external resource group
+	creds.ResourceGroup = credentials.ResourceGroupPrefix + uuid.NewUUID().String()
```
Comment (andyzhangx) [inline test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:77]: it's better use another rg var, use overwritten is not a good idea.

File: test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go [hunk 6]
```diff
@@ -46,15 +56,52 @@ func (t *DynamicallyProvisionedVolumeSnapshotTest) Run(client clientset.Interfac
...
-	ginkgo.By("creating volume snapshot class")
+	ginkgo.By("Checking Prow test resource group")
+	creds, err := credentials.CreateAzureCredentialFile(false)
+	framework.ExpectNoError(err, fmt.Sprintf("Error getting creds for AzurePublicCloud %v", err))
+
+	text := "Prow test resource group: " + creds.ResourceGroup
+	ginkgo.By(text)
+
+	//create external resource group
+	externalcreds, err := credentials.CreateAzureCredentialFile(false)
```
Comment (andyzhangx) [inline test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:73]: why create two `CreateAzureCredentialFile`? what about this:
```
externalRG := credentials.ResourceGroupPrefix + uuid.NewUUID().String()
azureClient, err := azure.GetAzureClient(creds.Cloud, creds.SubscriptionID, creds.AADClientID, creds.TenantID, creds.AADClientSecret)
...
_, err = azureClient.EnsureResourceGroup(ctx, externalRG, creds.Location, nil)
```

File: test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go [hunk 7]
```diff
@@ -46,15 +56,52 @@ func (t *DynamicallyProvisionedVolumeSnapshotTest) Run(client clientset.Interfac
...
+	framework.ExpectNoError(err, fmt.Sprintf("Error getting creds for AzurePublicCloud %v", err))
+	defer func() {
+		err := credentials.DeleteAzureCredentialFile()
+		framework.ExpectNoError(err)
+	}()
+	externalcreds.ResourceGroup = credentials.ResourceGroupPrefix + uuid.NewUUID().String()
+	text = "Creating external resource group: " + externalcreds.ResourceGroup
```
Comment (andyzhangx) [inline test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:80]: `ginkgo.By("Creating external resource group: " + externalcreds.ResourceGroup)`

File: test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go [hunk 8]
```diff
@@ -46,15 +56,52 @@ func (t *DynamicallyProvisionedVolumeSnapshotTest) Run(client clientset.Interfac
...
+	defer func() {
+		// Only delete resource group the test created
+		if strings.HasPrefix(externalcreds.ResourceGroup, credentials.ResourceGroupPrefix) {
+			e2elog.Logf("Deleting resource group %s", externalcreds.ResourceGroup)
+			err := azureClient.DeleteResourceGroup(ctx, externalcreds.ResourceGroup)
+			framework.ExpectNoError(err)
+		}
+	}()
+
+	ginkgo.By("creating volume snapshot class with external rg")
```
Comment (andyzhangx) [inline test/e2e/testsuites/dynamically_provisioned_volume_snapshot_tester.go:98]: print out the external rg name

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

**Pull request merged**